driver-core/src/main/com/mongodb/internal/connection/ConcurrentPool.java

```diff
@@ -136,17 +150,18 @@ public T get(final long timeout, final TimeUnit timeUnit) {
     }
 
     public void prune() {
-        int currentAvailableCount = getAvailableCount();
-        for (int numAttempts = 0; numAttempts < currentAvailableCount; numAttempts++) {
-            if (!acquirePermit(10, TimeUnit.MILLISECONDS)) {
-                break;
+        for (RemovalReportingIterator<T> iter = available.iterator(); iter.hasNext();) {
+            T cur = iter.next();
+            Prune shouldPrune = itemFactory.shouldPrune(cur);
+            if (shouldPrune == Prune.YES) {
+                boolean removed = iter.reportingRemove();
+                if (removed) {
+                    close(cur);
+                }
             }
-            T cur = available.pollFirst();
-            if (cur == null) {
-                releasePermit();
+            if (shouldPrune == Prune.STOP) {
```

Comment:
For readability I would put this check first.

Reply:
Done